source/extensions/http/header_validators/envoy_default/http2_header_validator.cc

```diff
@@ -410,30 +400,106 @@ Http2HeaderValidator::validateGenericHeaderName(const HeaderString& name) {
   return HeaderEntryValidationResult::success();
 }
 
-HeaderValidator::ValidationResult
+ValidationResult
 Http2HeaderValidator::validateRequestTrailers(const ::Envoy::Http::RequestTrailerMap& trailer_map) {
-  HeaderValidator::ValidationResult result = validateTrailers(trailer_map);
+  ValidationResult result = validateTrailers(trailer_map);
   if (!result.ok()) {
     stats_.incMessagingError();
   }
   return result;
 }
 
-HeaderValidator::TrailersTransformationResult
-Http2HeaderValidator::transformRequestTrailers(::Envoy::Http::RequestTrailerMap& trailer_map) {
+::Envoy::Http::HeaderValidator::TransformationResult
+ServerHttp2HeaderValidator::transformRequestTrailers(
+    ::Envoy::Http::RequestTrailerMap& trailer_map) {
   sanitizeHeadersWithUnderscores(trailer_map);
-  return TrailersTransformationResult::success();
+  return ::Envoy::Http::HeaderValidator::TransformationResult::success();
 }
 
-HeaderValidator::ValidationResult Http2HeaderValidator::validateResponseTrailers(
+ValidationResult Http2HeaderValidator::validateResponseTrailers(
     const ::Envoy::Http::ResponseTrailerMap& trailer_map) {
-  HeaderValidator::ValidationResult result = validateTrailers(trailer_map);
+  ValidationResult result = validateTrailers(trailer_map);
   if (!result.ok()) {
     stats_.incMessagingError();
   }
   return result;
 }
 
+::Envoy::Http::HeaderValidator::RequestHeadersTransformationResult
+ServerHttp2HeaderValidator::transformRequestHeaders(::Envoy::Http::RequestHeaderMap& header_map) {
+  sanitizeHeadersWithUnderscores(header_map);
+  if (!config_.uri_path_normalization_options().skip_path_normalization()) {
+    auto path_result = path_normalizer_.normalizePathUri(header_map);
+    if (!path_result.ok()) {
+      return path_result;
+    }
+  }
+
+  // Transform H/2 extended CONNECT to H/1 UPGRADE, so that request processing always observes H/1
+  // UPGRADE requests
+  if (::Envoy::Http::Utility::isH2UpgradeRequest(header_map)) {
+    ::Envoy::Http::Utility::transformUpgradeRequestFromH2toH1(header_map);
```

Comment:
I really don''t love that both paths return a transformation result with potential new headers but this changes headers inline.  I feel like it's a footgun waiting to happen.  I'd rather fix the APIs to be consistent
cc @mattklein123 for thoughts

Reply:
The main issue here is not so much with the API but with the violation of invariant that header map is not mutated by the upstream codec. Violation of this invariant breaks a number of places. For example, HCM encoder path breaks because it observed an upgrade request on the decode path and then extended CONNECT request when processing response. This also breaks retries and hedging to a mixed protocol pool, which now needs to transform  in both H/1 and H/2 codecs.

Changing it is doable but is a significant refactor. It is also unclear if there could any effect on filters that may keep state based on the type of request, which is changed when it is sent to the upstream.

We can tackle it, but it is a dangerous change.

Reply:
I honestly don't have any context on this ongoing work to understand the concern. If we aren't going to fix Alyssa's concerns now can we just add more comments/TODO to clearly spell them out? Thank you.

Reply:
yes, i think we can do that. Let me craft the PR to check this.

Reply:
ah, I think I was conflating RequestHeaderValidationResult with ResponseHeaderValidationResult.

sigh.
OK no objections to the API then, will review